### PR TITLE
Expand FA coverage to padding masks

### PR DESCRIPTION
**Pull request by ksivaman**

### CODE REVIEW

File: tests/pytorch/test_sanity.py [hunk 1]
```diff
@@ -256,6 +256,29 @@ def _test_sanity_e2e(block, bs, dtype, config, fp8_recipe, skip_wgrad):
...
+    te_inp_hidden_states = torch.randn(
+        config.seq_len, bs, config.hidden_size, dtype=dtype, requires_grad=True
+    ).cuda()
+
+    if dtype == torch.float32:
+        mask_shape = torch.Size([1, 1, config.seq_len, config.seq_len])
+    else:
+        mask_shape = torch.Size([config.seq_len, bs])
+
+    te_inp_attn_mask = torch.rand(mask_shape).cuda().bool()
```
Comment (timmoon10) [inline tests/pytorch/test_sanity.py:269]: Why does the FP32 test require a 4D mask while other dtypes use a 2D mask?

File: qa/L0_lint/pylintrc [hunk 2]
```diff
@@ -22,6 +22,7 @@ disable=too-many-locals,
         attribute-defined-outside-init,
         global-statement,
         too-many-branches,
+        too-many-boolean-expressions,
```
Comment (timmoon10) [inline qa/L0_lint/pylintrc:25]: I presume this is to suppress a warning from https://github.com/NVIDIA/TransformerEngine/blob/a89e43c2c48023cfa539d62f6e351cc6b486ce60/transformer_engine/pytorch/attention.py#L562

I'd prefer keeping as many PyLint warnings as possible to keep our style disciplined. The offending if-statement seems sensible to me, so how about we locally suppress the warning instead with `# pylint: disable=too-many-boolean-expressions`.

File: tests/pytorch/test_sanity.py (hunk 1 above)
Comment (ksivaman) [reply tests/pytorch/test_sanity.py:269]: The FP32 doesn't take the FA path and so uses the PyTorch torch softmax path for padding mask which expects this

Comment (timmoon10) [reply tests/pytorch/test_sanity.py:269]: Ah, so FA vs PyT is no longer an implementation detail since it expects a different mask format. This makes me think we need an option to explicitly enable or disable FA, and to error out instead of falling back to PyT.

**Review by timmoon10 — CHANGES_REQUESTED** [review]

File: tests/pytorch/test_sanity.py (hunk 1 above)
Comment (timmoon10) [reply tests/pytorch/test_sanity.py:269]: Alternatively, we should expect the mask to be in the format for PyT, and then convert it to the FA format internally so it isn't visible to the user.

File: transformer_engine/pytorch/attention.py [hunk 3]
```diff
@@ -41,6 +42,19 @@
 __all__ = ["DotProductAttention"]
 
 
+@jit_fuser
+def get_cu_seqlens(padding_mask: torch.Tensor) -> torch.Tensor:
+    """
+    Given a padding mask of shape [seq_len, batch_size], returns an int32
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:48]: I guess by `seq_len` you mean `max_seqlen`? Also, should the mask tensor be required to be a CUDA tensor?

File: transformer_engine/pytorch/attention.py [hunk 4]
```diff
@@ -355,18 +369,30 @@ def forward(
...
-        cu_seqlens = torch.arange(
-            0,
-            (batch_size + 1) * seqlen,
-            step=seqlen,
-            dtype=torch.int32,
-            device=query_layer.device)
+        if self.attn_mask_type == "padding":
+            assert (
+                attention_mask is not None
+            ), "Boolean attention mask must be provided for padding."
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:375]: So we're settled on requiring a mask tensor than a cu_seqlen tensor from users?

File: transformer_engine/pytorch/attention.py [hunk 5]
```diff
@@ -529,36 +555,25 @@ def forward(
         """
 
         use_flash_attention = self.use_flash_attention
-        if (query_layer.dtype not in [torch.bfloat16, torch.float16]
+        if (query_layer.dtype not in [torch.bfloat16, torch.float16] # pylint: disable=too-many-boolean-expressions
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:558]: Could probably use `any(x.dtype not in [torch.bfloat16, torch.float16] for x in [query_layer, key_layer, value_layer])`. I think pylint recognizes this as one boolean expression.

File: transformer_engine/pytorch/attention.py [hunk 6]
```diff
@@ -529,36 +555,25 @@ def forward(
...
-            use_flash_attention = False
-
-        if use_flash_attention:
-            if checkpoint_core_attention:
-                return self._checkpointed_attention_forward(self.flash_attention,
-                                                            query_layer,
-                                                            key_layer,
-                                                            value_layer)
-            return self.flash_attention(query_layer, key_layer, value_layer)
+        attention_func = self.flash_attention if use_flash_attention else self.unfused_attention
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:566]: Should we include self.fused_attention in this logic as well? In my PR, I kept the selection order as flash -> fused -> unfused.

File: transformer_engine/pytorch/attention.py [hunk 7]
```diff
@@ -756,10 +771,14 @@ def forward(
         """MultiHeadAttention FWD"""
         # hidden_states: [sq, b, h]
 
-        if self.attn_mask_type != "causal" and attention_mask is not None:
+        if self.attn_mask_type == "padding":
+            assert (
+                attention_mask is not None
+            ), "Boolean attention mask must be provided for padding."
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:777]: The m512 version of the fused attention supports padding as well, but it doesn't require a mask tensor. Unfortunately which backend to use for fused attention is only determined later in DPA forward. So maybe we can move this check to the FlashAttention module instead.

Comment (ksivaman) [reply transformer_engine/pytorch/attention.py:777]: How does padding work without a mask tensor provided?

File: transformer_engine/pytorch/attention.py (hunk 6 above)
Comment (ksivaman) [reply transformer_engine/pytorch/attention.py:566]: Yes I like this hierarchy. I'm switching to it as I resolve all the conflicts

File: transformer_engine/pytorch/attention.py (hunk 4 above)
Comment (ksivaman) [reply transformer_engine/pytorch/attention.py:409]: Yeah I think so

File: transformer_engine/pytorch/attention.py (hunk 3 above)
Comment (ksivaman) [reply transformer_engine/pytorch/attention.py:59]: Yeah it is the `max_seqlen`.

Comment (ksivaman) [reply transformer_engine/pytorch/attention.py:59]: Yeah it should be cuda tensor since the inp is cuda as well.

File: tests/pytorch/test_sanity.py (hunk 1 above)
Comment (ptrendx) [reply tests/pytorch/test_sanity.py:233]: Agree with Tim. OR, accept both in both cases and generate the right one for the current implementation.

To be honest though, the current API for mask type is just bad. For example, what if you want padding AND causal? Then your only option is set the padding mask type and create the causal mask yourself (which would mean you can't do FA in the current implementation which is bad). There is also nothing stopping you really from having arbitrary padding mask (as in, with random elements zeroed out), which would break the assumptions you have in this PR. I think what we could do is introduce new names for the padding type (let's say "pad", "arbitrary" and "no_mask"), deprecate the old names and add causal as separate switch. Then we could say that pad only accepts the "mask" that is actually just list of sequence lengths, arbitrary is whatever you want (and will not go through FA) and causal could be switched irrespective of the mask type.

**Comment by ksivaman** [comment]
Closing in favor of #302
